File: nodejs/service/comment.service.js
const commentModel = require("../model/comment.model");

class CommentService {
    addComment = async (req) => {
        const { comment, postId } = req.body;
        if (!comment) {
            throw new Error("Comment is required");
        }
        return await commentModel.create({
            comment,
            postId,
            userId: req.user.id,
        });
    };
    getComments = async (req) => {
        return await commentModel
            .find({ postId: req.params.postId })
            .populate("userId", "username")
            .populate({
                path: "postId",
                select: "userId",
            })
            .sort({ createdAt: -1 });
    };
    updateComment = async (req) => {
        const comment = await commentModel
            .findById(req.params.id)
            .populate("postId");
        if (!comment) {
            throw new Error("Comment not found");
        }
        const isCommentOwner =
            comment.userId.toString() === req.user.id;
        const isPostOwner =
            comment.postId.userId.toString() === req.user.id;
        if (!isCommentOwner && !isPostOwner) {
            throw new Error("Unauthorized");
        }
        comment.comment = req.body.comment;
        await comment.save();
        return comment;
    };
    deleteComment = async (req) => {
        const comment = await commentModel
            .findById(req.params.id)
            .populate("postId");
        if (!comment) {
            throw new Error("Comment not found");
        }
        const isCommentOwner =
            comment.userId.toString() === req.user.id;
        const isPostOwner =
            comment.postId.userId.toString() === req.user.id;
        if (!isCommentOwner && !isPostOwner) {
            throw new Error("Unauthorized");
        }
        await comment.deleteOne();
        return {
            message: "Comment deleted successfully",
        };
    };
}
module.exports = new CommentService();

File: DJANGO/api/services/comment_service.py
from api.models import Comment, Post

class CommentService:

    @staticmethod
    def add_comment(data, user):

        if not data.get("comment"):
            raise Exception("Comment is required")
        post = Post.objects.get(id=data["post"])
        comment = Comment.objects.create(
            comment=data["comment"],
            user=user,
            post=post
        )

        return comment

    @staticmethod
    def get_comments(post_id):
        return Comment.objects.filter(
            post_id=post_id
        ).select_related("user").order_by("-createdAt")

    @staticmethod
    def delete_comment(id, user):

        comment = Comment.objects.select_related("post").get(id=id)

        is_comment_owner = comment.user.id == user.id
        is_post_owner = comment.post.user.id == user.id

        if not is_comment_owner and not is_post_owner:
            raise Exception("Unauthorized")

        comment.delete()

    @staticmethod
    def update_comment(id, data, user):

        comment = Comment.objects.select_related("post").get(id=id)

        is_comment_owner = comment.user.id == user.id
        is_post_owner = comment.post.user.id == user.id

        if not is_comment_owner and not is_post_owner:
            raise Exception("Unauthorized")
        
        if not data.get("comment"):
            raise Exception("Comment is required")

        comment.comment = data["comment"]
        comment.save()

        return comment